ceph.md

```diff
@@ -0,0 +1,245 @@
+
+The commands in this section should be run on a Ceph server.
+
+Create pools for Nova (vms), Cinder (volumes) and Glance (images).
+```
+for P in vms volumes images; do 
+  cephadm shell -- ceph osd pool create $P;
+  cephadm shell -- ceph osd pool application enable $P rbd;
+done
+```
+
+Create a cephx key which OpenStack can use to access the pools.
+```
+cephadm shell -- \
+   ceph auth add client.openstack \ 
+     mgr 'allow *' \
+	 mon 'allow r' \
+	 osd 'allow class-read object_prefix rbd_children, allow rwx pool=vms, allow rwx pool=volumes, allow rwx pool=images'
+```
+
+Export the cephx key and Ceph configuration file.
+```
+cephadm shell -- ceph auth get client.openstack > /etc/ceph/ceph.client.openstack.keyring
+cephadm shell -- ceph config generate-minimal-conf > /etc/ceph/ceph.conf
+```
+
+## Create a Ceph Secret
+
+Transfer the cephx key and Ceph configuration file from the previous
+section to a host which can create resources in the openstack
+namespace. Base64 encode these files and store them in two variables.
+```
+KEY=$(cat /etc/ceph/ceph.client.openstack.keyring | base64 -w 0)
+CONF=$(cat /etc/ceph/ceph.conf | base64 -w 0)
+```
+Use the variables to create a `ceph-conf-files` secret.
+```
+cat <<EOF > ceph_secret.yaml
+apiVersion: v1
+data:
+  ceph.client.openstack.keyring: $KEY
+  ceph.conf: $CONF
+kind: Secret
+metadata:
+  name: ceph-conf-files
+  namespace: openstack
+type: Opaque
+EOF
+
+oc create -f ceph_secret.yaml
+```
+
+The Ceph FSID can be extracted from the secret using the following
+command.
+```
+FSID=$(oc get secret ceph-conf-files -o json | jq -r '.data."ceph.conf"' | base64 -d | grep fsid | sed -e 's/fsid = //')
+```
+The above will be useful for configuration snippets covered later in
+the document.
+
+## Access the Ceph Secret via extraMounts
+
+Use
+[extraMounts](https://github.com/openstack-k8s-operators/lib-common/commit/c3752c8a43423c755adb16aebf644c304befc072)
+in CRs for pods which need to access the Ceph
+secret. For example, the sample
+[core_v1beta1_openstackcontrolplane_network_isolation_ceph.yaml](https://github.com/openstack-k8s-operators/openstack-operator/blob/main/config/samples/core_v1beta1_openstackcontrolplane_network_isolation_ceph.yaml)
+has the following in the `OpenStackControlPlane` CR so that the
+Glance and Cinder volume pods mount the cephx key and Ceph
+configuration files in /etc/ceph.
+
+```
+apiVersion: core.openstack.org/v1beta1
+kind: OpenStackControlPlane
+spec:
+  extraMounts:
+    - name: v1
+      region: r1
+      extraVol:
+        - propagation:
+          - CinderVolume
+          - GlanceAPI
+          extraVolType: Ceph
+          volumes:
+          - name: ceph
+            projected:
+              sources:
+              - secret:
+                  name: ceph-conf-files
+          mounts:
+          - name: ceph
+            mountPath: "/etc/ceph"
+            readOnly: true
+```
+The `OpenStackDataPlane` can also use `extraMounts`.
```

Comment:
this section feels a bit handwavy


im conficlitd between explaining just watch is required an providing a complete example.

perhaps we could provide examples of a complete
OpenStackControlPlane cr and OpenStackDataPlane/OpenStackDataPlaneRole/OpenStackDataPlaneNode crs

at the end

so we can discusee the snipits here in each section but then show what it looks like when its combined.

reition now it could be assume that each sech is used standalone as separate cr when infact these will be combined.

Reply:
> im conficlitd between explaining just [what] is required an providing a complete example.
> 
> perhaps we could provide examples of a complete OpenStackControlPlane cr and OpenStackDataPlane/OpenStackDataPlaneRole/OpenStackDataPlaneNode crs
> 
> at the end

I link to a complete example of `OpenStackControlPlane` node on lines 98, 158 and 185.
I can link to a complete example of `OpenStackDataPlane` too.

> so we can discusee the snipits here in each section but then show what it looks like when its combined.

Yes, I did want to have a single document which focussed on the necessary snippets required only for Ceph.

> reition now it could be assume that each sech is used standalone as separate cr when infact these will be combined.

I thought the including the `kind` of each CR it would be clear which CRs are separate and which could be combined. One of my pre-reqs on line 16 is that the reader knows to use the openstack-k8s-operators project and is looking to modify it to include Ceph. Thus, I focussed only on the diff but also provided an explanation.

Reply:
I'll update the next PS to spell it out more and link to complete examples.

Reply:
A link to the example is sufficient IMO, the whole `OpenstackControlPlane` CR is too verbose and should also be updated in case anything change, so I would stay on the short version, maybe with some "dots" in the middle, so the reader knows we're referring to something we can't just `oc apply -i`

Reply:
@SeanMooney I've added a section at the bottom called "Full Exampes" which I think points out what you are asking for. It links to the sample ControlPlane CR with Ceph config under the openstack-operator. The data-plane operator didn't have a ceph example so I created one in another PR and linked to it from the latest version of the document.

https://github.com/openstack-k8s-operators/dataplane-operator/pull/216

Reply:
thats no quite what i was askign for but it will do for now.

i was asking for you to embed it directly in the doc but i think long term we can have a better solution by maintaining a repo with example that are tested so linking for now i guess is ok.

thanks fro creating the datapalne-oeprator example